numpyro/distributions/transforms.py

```diff
@@ -92,10 +96,24 @@ def codomain(self):
         elif self.domain is constraints.real_vector:
             return constraints.real_vector
         elif isinstance(self.domain, constraints.greater_than):
-            return constraints.greater_than(self.__call__(self.domain.lower_bound))
+            if not_jax_tracer(self.scale) and jnp.all(self.scale < 0):
+                return constraints.less_than(self(self.domain.lower_bound))
+            # we suppose scale > 0 for any tracer
+            else:
+                return constraints.greater_than(self(self.domain.lower_bound))
+        elif isinstance(self.domain, constraints.less_than):
+            if not_jax_tracer(self.scale) and jnp.all(self.scale < 0):
+                return constraints.greater_than(self(self.domain.upper_bound))
+            # we suppose scale > 0 for any tracer
+            else:
+                return constraints.less_than(self(self.domain.upper_bound))
```

Comment:
I like that in NumPyro you added a customizable `domain` arg. That might also be helpful in PyTorch, especially for `TransformedDistribution`. When we discussed it in the past we decided that the constraint transform logic would be overly complex.

Indeed the logic here is getting quite complex, e.g. you need to make a `> 0` assumption. Do you think it would be simpler to add a customizable `codomain` arg as well? Maybe similarly `TransformedDistribution` should take a `.support` arg in case it cannot be computed directly. I know I've created subclasses of `TransformedDistribution` merely to add a proper `.support`.

Reply:
> created subclasses of TransformedDistribution merely to add a proper .support

I also faced this issue in the past. I agree that having arg `support` would be simpler in situations such as `TransformedDist(LogNormal(), ExpTransform())` because `codomain` of `ExpTransform` is fixed to `positive`. But I am also happy with the logic
```
    @property
    def support(self):
        domain = self.base_dist.support
        for t in self.transforms:
            t.domain = domain
            domain = t.codomain
        return domain
```
in `TransformedDistribution`, which should cover most usage cases.

Similarly, having a customizable `codomain` is convenient but we still need this logic in case users didn't provide `codomain`. IIRC, we have `domain` arg here so that users can provide a correct sequence of `domain/codomain` in TransformedDistribution in the constructor (for this, having `codomain` arg, as you said, would be much simpler), but later, we used the above logic to compute `support` because it is nontrivial for users to handle that `domain/codomain` sequence.

Reply:
Thanks for explaining!